routing/router.go

```diff
@@ -3254,50 +3259,102 @@ func backwardPass(unifiers []*edgeUnifier, useMinAmt bool,
 			}
 		}
 
-		// Get an edge for the specific amount that we want to forward.
-		edge := edgeUnifier.getEdge(runningAmt, bandwidthHints, 0)
-		if edge == nil {
-			log.Errorf("Cannot find policy with amt=%v for hop %v",
-				runningAmt, i)
+		if i == len(unifiers)-1 {
+			// Get an edge for the specific amount that we want to
+			// forward.
+			edge = edgeUnifier.getEdge(
+				runningAmt, bandwidthHints, 0,
+			)
+			if edge == nil {
+				log.Errorf("Cannot find policy with amt=%v "+
+					"for hop %v", runningAmt, i)
 
-			return nil, 0, ErrNoChannel{
-				position: i,
+				return nil, 0, ErrNoChannel{
+					position: i,
+				}
 			}
-		}
 
-		// Add fee for this hop.
-		if !localChan {
-			runningAmt += edge.policy.ComputeFee(runningAmt)
+			// The last hop only forwards the amount that the
+			// destination should receive.
+			fee = 0
+		} else {
+			// The amount that the current hop needs to forward is
+			// equal to the incoming amount of the following hop
+			// that we already dealt with.
+			runningAmt = nextIncoming
+
+			// The fee that needs to be paid to the current node is
+			// based on the amount that this node needs to forward
+			// and its policy for its outgoing channel. This policy
+			// is stored as part of the incoming channel of the
+			// following hop.
+			outboundFee := unifiedEdges[i+1].policy.ComputeFee(
+				runningAmt,
+			)
+
+			netAmount := runningAmt + outboundFee
+
+			// The unified edge is selected such that it can support
+			// forwarding the total amount, including the inbound
+			// fee. The previous hop's outbound fee is also supplied
+			// such that the inbound fee can be capped.
+			edge = edgeUnifier.getEdge(
+				netAmount, bandwidthHints, outboundFee,
+			)
+			if edge == nil {
+				return nil, 0, ErrNoChannel{
+					position: i,
+				}
+			}
+
+			log.Tracef("Select channel %v at position %v",
+				edge.policy.ChannelID, i)
+
+			fee = int64(outboundFee)
```

Comment:
It looks like you are converting to int64 only to convert immediately back to MSats.